### PR TITLE
Adafruit uart bluetooth support

### PR DESCRIPTION
**Pull request by kanru**
## Description

Extracted from https://github.com/qmk/qmk_firmware/pull/11011
This PR adds bluetooth support via Adafruit UART Friend UART interface

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by fauxpark** [comment]
I've finally figured out why the SPI Friend was not working on ARM - turns out there is a bug in spi_master: https://github.com/qmk/qmk_firmware/pull/11404
So I am now able to progress with the Bluetooth refactor and API. It should be a lot nicer to implement; there is a standard set of functions each driver can implement (or not), Bluetooth drivers will live in `drivers/bluetooth`, and rules.mk configuration more like RGB Matrix:

```makefile
BLUETOOTH_ENABLE = yes
BLUETOOTH_DRIVER = BluefruitLE_SPI # (, BluefruitLE_UART, RN42)
```

File: tmk_core/protocol/lufa/adafruit_ble_uart.c [hunk 1]
```diff
@@ -0,0 +1,377 @@
+#include "adafruit_ble.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <alloca.h>
+#include "protocol/serial.h"
+#include "debug.h"
+#include "timer.h"
```
Comment (yoyostile) [inline tmk_core/protocol/lufa/adafruit_ble_uart.c:8]:
```suggestion
#include "timer.h"
#include "progmem.h"
```

File: tmk_core/protocol/lufa/adafruit_ble_uart.c [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+            }
+            snprintf_P(cmdbuf, sizeof(cmdbuf), PSTR("AT+BLEHIDMOUSEBUTTON=%d"), item->mousemove.buttons);
+            return at_command(cmdbuf, NULL, 0);
+#endif
+        default:
+            return true;
+    }
+}
+
+bool adafruit_ble_send_keys(uint8_t hid_modifier_mask, uint8_t *keys, uint8_t nkeys) {
```
Comment (yoyostile) [inline tmk_core/protocol/lufa/adafruit_ble_uart.c:289]:
```suggestion
void adafruit_ble_send_keys(uint8_t hid_modifier_mask, uint8_t *keys, uint8_t nkeys) {
```

File: tmk_core/protocol/lufa/adafruit_ble_uart.c [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+        }
+
+        nkeys -= 6;
+        keys += 6;
+    }
+
+    return true;
+}
+
+bool adafruit_ble_send_consumer_key(uint16_t keycode, int hold_duration) {
```
Comment (yoyostile) [inline tmk_core/protocol/lufa/adafruit_ble_uart.c:318]:
```suggestion
void adafruit_ble_send_consumer_key(uint16_t keycode) {
```

File: tmk_core/protocol/lufa/adafruit_ble_uart.c [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+    struct queue_item item;
+
+    item.queue_type = QTConsumer;
+    item.consumer   = keycode;
+
+    return enqueue(&send_queue, &item);
+}
+
+#ifdef MOUSE_ENABLE
+bool adafruit_ble_send_mouse_move(int8_t x, int8_t y, int8_t scroll, int8_t pan, uint8_t buttons) {
```
Comment (yoyostile) [inline tmk_core/protocol/lufa/adafruit_ble_uart.c:328]:
```suggestion
void adafruit_ble_send_mouse_move(int8_t x, int8_t y, int8_t scroll, int8_t pan, uint8_t buttons) {
```

File: tmk_core/protocol/lufa/adafruit_ble_uart.c [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+    while (nkeys >= 0) {
+        item.key.keys[0] = keys[0];
+        item.key.keys[1] = nkeys >= 1 ? keys[1] : 0;
+        item.key.keys[2] = nkeys >= 2 ? keys[2] : 0;
+        item.key.keys[3] = nkeys >= 3 ? keys[3] : 0;
+        item.key.keys[4] = nkeys >= 4 ? keys[4] : 0;
+        item.key.keys[5] = nkeys >= 5 ? keys[5] : 0;
+
+        if (!enqueue(&send_queue, &item)) {
+            return false;
```
Comment (yoyostile) [inline tmk_core/protocol/lufa/adafruit_ble_uart.c:304]:
```suggestion
            return;
```

File: tmk_core/protocol/lufa/adafruit_ble_uart.c [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+        }
+
+        if (nkeys <= 6) {
+            return true;
```
Comment (yoyostile) [inline tmk_core/protocol/lufa/adafruit_ble_uart.c:308]:
```suggestion
            return;
```

File: tmk_core/protocol/lufa/adafruit_ble_uart.c [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+        }
+
+        nkeys -= 6;
+        keys += 6;
+    }
+
+    return true;
```
Comment (yoyostile) [inline tmk_core/protocol/lufa/adafruit_ble_uart.c:315]:
```suggestion
    return;
```

File: tmk_core/protocol/lufa/adafruit_ble_uart.c [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+}
+
+bool adafruit_ble_send_consumer_key(uint16_t keycode, int hold_duration) {
+    struct queue_item item;
+
+    item.queue_type = QTConsumer;
+    item.consumer   = keycode;
+
+    return enqueue(&send_queue, &item);
```
Comment (yoyostile) [inline tmk_core/protocol/lufa/adafruit_ble_uart.c:324]:
```suggestion
    enqueue(&send_queue, &item);
```

File: tmk_core/protocol/lufa/adafruit_ble_uart.c [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+    struct queue_item item;
+
+    item.queue_type        = QTMouseMove;
+    item.mousemove.x       = x;
+    item.mousemove.y       = y;
+    item.mousemove.scroll  = scroll;
+    item.mousemove.pan     = pan;
+    item.mousemove.buttons = buttons;
+
+    return enqueue(&send_queue, &item);
```
Comment (yoyostile) [inline tmk_core/protocol/lufa/adafruit_ble_uart.c:338]:
```suggestion
    enqueue(&send_queue, &item);
```

**Review by yoyostile — COMMENTED** [review]
Imho necessary because of https://github.com/qmk/qmk_firmware/pull/11556. 
Now everything works fine again locally.

**Comment by kanru** [comment]
@yoyostile nice, I'll test soon. What board are you using? Does the bluetooth work well for you?

**Comment by yoyostile** [comment]
@kanru hhkb ble mod from ydkb. Bluetooth works just fine!

File: tmk_core/protocol/lufa/adafruit_ble_uart.c [hunk 10]
```diff
@@ -0,0 +1,378 @@
+#include "adafruit_ble.h"
```
Comment (drashna) [inline tmk_core/protocol/lufa/adafruit_ble_uart.c:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

**Review by drashna — CHANGES_REQUESTED** [review]

File: tmk_core/protocol/lufa/adafruit_ble_uart.c (hunk 10 above)
Comment (kanru) [reply tmk_core/protocol/lufa/adafruit_ble_uart.c:17]: Done!

**Review by drashna — APPROVED** [review]

File: tmk_core/protocol/lufa/adafruit_ble_uart.c [hunk 11]
```diff
@@ -0,0 +1,394 @@
...
+        struct __attribute__((packed)) {
+            uint8_t modifier;
+            uint8_t keys[6];
+        } key;
+        uint16_t consumer;
+#ifdef MOUSE_ENABLE
+        struct __attribute__((packed)) {
+            int8_t  x, y, scroll, pan;
+            uint8_t buttons;
+        } mousemove;
```
Comment (drashna) [inline tmk_core/protocol/lufa/adafruit_ble_uart.c:71]: You could probably use the `report_mouse_t` structure here, instead.

**Pull request closed without merging**